nixos/tests/ceph-multi-node.nix

```diff
@@ -0,0 +1,231 @@
+    osd0 = generateHost { pkgs = pkgs; cephConfig = cephConfigOsd0; networkConfig = networkOsd0; };
+    osd1 = generateHost { pkgs = pkgs; cephConfig = cephConfigOsd1; networkConfig = networkOsd1; };
+  };
+
+  testScript = { ... }: ''
+    startAll;
+
+    $monA->waitForUnit("network.target");
+    $osd0->waitForUnit("network.target");
+    $osd1->waitForUnit("network.target");
+
+    # Create the ceph-related directories
+    $monA->mustSucceed(
+      "mkdir -p /var/lib/ceph/mgr/ceph-a",
+      "mkdir -p /var/lib/ceph/mon/ceph-a",
+      "chown ceph:ceph -R /var/lib/ceph/",
+      "mkdir -p /etc/ceph",
+      "chown ceph:ceph -R /etc/ceph"
+    );
+    $osd0->mustSucceed(
+      "mkdir -p /var/lib/ceph/osd/ceph-0",
+      "chown ceph:ceph -R /var/lib/ceph/",
+      "mkdir -p /etc/ceph",
+      "chown ceph:ceph -R /etc/ceph"
+    );
+    $osd1->mustSucceed(
+      "mkdir -p /var/lib/ceph/osd/ceph-1",
+      "chown ceph:ceph -R /var/lib/ceph/",
+      "mkdir -p /etc/ceph",
+      "chown ceph:ceph -R /etc/ceph"
+    );
+
+    # Bootstrap ceph-mon daemon
+    $monA->mustSucceed(
+      "sudo -u ceph ceph-authtool --create-keyring /tmp/ceph.mon.keyring --gen-key -n mon. --cap mon 'allow *'",
+      "sudo -u ceph ceph-authtool --create-keyring /etc/ceph/ceph.client.admin.keyring --gen-key -n client.admin --cap mon 'allow *' --cap osd 'allow *' --cap mds 'allow *' --cap mgr 'allow *'",
+      "sudo -u ceph ceph-authtool /tmp/ceph.mon.keyring --import-keyring /etc/ceph/ceph.client.admin.keyring",
+      "monmaptool --create --add a 192.168.1.1 --fsid 066ae264-2a5d-4729-8001-6ad265f50b03 /tmp/monmap",
+      "sudo -u ceph ceph-mon --mkfs -i a --monmap /tmp/monmap --keyring /tmp/ceph.mon.keyring",
+      "sudo -u ceph touch /var/lib/ceph/mon/ceph-a/done",
+      "systemctl start ceph-mon-a"
+    );
+    $monA->waitForUnit("ceph-mon-a");
+    $monA->mustSucceed("ceph mon enable-msgr2");
+
+    # Can't check ceph status until a mon is up
+    $monA->succeed("ceph -s | grep 'mon: 1 daemons'");
+
+    # Start the ceph-mgr daemon, it has no deps and hardly any setup
+    $monA->mustSucceed(
+      "ceph auth get-or-create mgr.a mon 'allow profile mgr' osd 'allow *' mds 'allow *' > /var/lib/ceph/mgr/ceph-a/keyring",
+      "systemctl start ceph-mgr-a"
+    );
+    $monA->waitForUnit("ceph-mgr-a");
+    $monA->waitUntilSucceeds("ceph -s | grep 'quorum a'");
+    $monA->waitUntilSucceeds("ceph -s | grep 'mgr: a(active,'");
+
+    # Send the admin keyring to the OSD machines
+    $osd0->mustSucceed("nc -vlkN 6800 > /etc/ceph/ceph.client.admin.keyring &");
+    $osd1->mustSucceed("nc -vlkN 6800 > /etc/ceph/ceph.client.admin.keyring &");
+    $osd0->waitForOpenPort("6800");
+    $osd1->waitForOpenPort("6800");
+    $monA->mustSucceed(
+      "nc 192.168.1.2 6800 < /etc/ceph/ceph.client.admin.keyring",
+      "nc 192.168.1.3 6800 < /etc/ceph/ceph.client.admin.keyring"
+    );
```

Comment:
The test framework sets up a shared folder in /tmp/shared that seems simpler to use.

Reply:
Oh, I didn't know that, I tried to find in the manual on how you can share stuff between the machine but failed to find it. Then its much better to use.

Reply:
Regarding the UUID and such, I would gladly discuss how we would do it. I wasn't sure if %{thing} works in the testScript part, but if it does, it would be easy to at least sanitize the testScript part.

Reply:
It definitely does! And it's not a major thing at all. :)

Reply:
Fixed to use a single attrSet as authoritative for all the IPs, UUIDs and such in 47604c7